File: backend/solver/pulse/solver/variants/leduc.py
"""Leduc Hold'em — 6-card, 2-street poker game for multi-round CFR validation.

Rules:
- 6-card deck: J, Q, K with 2 suits each
- Each player antes 1 chip, gets 1 private card
- Street 1: betting round (check/bet(2))
- 1 community card dealt
- Street 2: betting round (check/bet(4))
- Showdown: pair (private matches board) beats high card; higher rank wins ties
- Max 2 bets per street (bet + raise)
"""

from __future__ import annotations

from itertools import permutations
from typing import Sequence

from ..cards import card, card_rank, card_suit
from ..game_tree import (
    Action, ActionInfo, ActionNode, ChanceNode, GameNode,
    TerminalNode, TreeBuilder,
)

# Leduc uses J=9, Q=10, K=11 with suits 0 and 1
LEDUC_RANKS = [9, 10, 11]
LEDUC_SUITS = [0, 1]
LEDUC_CARDS = [card(r, s) for r in LEDUC_RANKS for s in LEDUC_SUITS]
LEDUC_RANK_NAMES = {9: "J", 10: "Q", 11: "K"}

ANTE = 1
BET_STREET1 = 2
BET_STREET2 = 4
MAX_BETS_PER_STREET = 2


class LeducTreeBuilder(TreeBuilder):
    """Builds the Leduc Hold'em game tree."""

    def build_tree(self) -> GameNode:
        root = ChanceNode(depth=0)

        # Deal 2 private cards (all ordered pairs from 6 cards)
        deals = [(c1, c2) for c1 in LEDUC_CARDS for c2 in LEDUC_CARDS if c1 != c2]
        prob = 1.0 / len(deals)  # 1/30

        for p0_card, p1_card in deals:
            label = f"{_cname(p0_card)}{_cname(p1_card)}"
            child = self._build_street1(p0_card, p1_card, depth=1)
            child.parent = root
            root.outcomes[label] = (prob, child)

        return root

    def _build_street1(self, p0_card: int, p1_card: int, depth: int) -> GameNode:
        """Build the first street betting tree."""
        pot = 2 * ANTE  # both players ante
        return self._build_betting_round(
            p0_card, p1_card, board_card=None,
            pot=pot, street=1, bet_size=BET_STREET1,
            history="", num_bets=0, depth=depth,
        )

    def _build_betting_round(self, p0_card: int, p1_card: int,
                             board_card: int | None,
                             pot: float, street: int, bet_size: int,
                             history: str, num_bets: int,
                             depth: int) -> GameNode:
        """Recursively build a betting round."""
        # Determine acting player
        # In each street, P0 acts first, then alternates
        actions_in_round = len(history.split("/")[-1]) if "/" in history else len(history)
        player = actions_in_round % 2

        player_card = p0_card if player == 0 else p1_card
        info_key = self.get_info_set_key(
            player, [player_card], history,
            board=[board_card] if board_card is not None else [],
        )

        # Terminal conditions within round
        if self._is_round_terminal(history):
            if history.endswith("f"):
                # Last action was fold
                folder = player  # current player folded? No — the player who folded
                # Actually, the fold is the last action taken by the previous player
                prev_player = 1 - player
                # The previous player folded
                winner = player  # current player wins
                payoff = pot / 2  # winner gets opponent's contribution
                payoffs = (payoff, -payoff) if winner == 0 else (-payoff, payoff)
                return TerminalNode(depth=depth, payoffs=payoffs, pot=pot)

            # Round ended with check-check or bet-call
            if street == 1:
                # Deal community card → go to street 2
                return self._deal_board_card(p0_card, p1_card, pot, history, depth)
            else:
                # Street 2 over → showdown
                return self._showdown(p0_card, p1_card, board_card, pot, depth)

        # Build action node
        actions = self._get_actions(history, num_bets, bet_size)
        node = ActionNode(
            depth=depth,
            player=player,
            actions=actions,
            info_set_key=info_key,
            pot=pot,
        )

        for action in actions:
            if action.label == "fold":
                new_hist = history + "f"
                new_pot = pot
                new_bets = num_bets
            elif action.label == "check":
                new_hist = history + "c"
                new_pot = pot
                new_bets = num_bets
            elif action.label.startswith("bet") or action.label.startswith("raise"):
                new_hist = history + "b"
                new_pot = pot + bet_size
                new_bets = num_bets + 1
            elif action.label == "call":
                new_hist = history + "k"  # 'k' for call to avoid ambiguity
                new_pot = pot + bet_size
                new_bets = num_bets
            else:
                raise ValueError(f"Unknown action: {action.label}")

            child = self._build_betting_round(
                p0_card, p1_card, board_card,
                new_pot, street, bet_size,
                new_hist, new_bets, depth + 1,
            )
            child.parent = node
            node.children[action.label] = child

        return node

    def _is_round_terminal(self, history: str) -> bool:
        """Check if the current betting round is over."""
        # Extract actions in current street
        current = history.split("/")[-1] if "/" in history else history
        if len(current) < 2:
            return False
        last_two = current[-2:]
        # Round ends on: cc (check-check), bk (bet-call), bf (bet-fold),
        # bbk (bet-raise-call), bbf (bet-raise-fold)
        if last_two in ("cc", "bk", "bf"):
            return True
        return False

    def _get_actions(self, history: str, num_bets: int, bet_size: int) -> list[ActionInfo]:
        """Get available actions based on history."""
        current = history.split("/")[-1] if "/" in history else history
        last_action = current[-1] if current else ""

        if last_action == "b":
            # Facing a bet: fold, call, or raise (if under max bets)
            actions = [
                ActionInfo(Action.FOLD, label="fold"),
                ActionInfo(Action.CALL, amount=bet_size, label="call"),
            ]
            if num_bets < MAX_BETS_PER_STREET:
                actions.append(ActionInfo(Action.RAISE, amount=bet_size, label="raise"))
            return actions
        else:
            # No bet facing: check or bet
            return [
                ActionInfo(Action.CHECK, label="check"),
                ActionInfo(Action.BET, amount=bet_size, label=f"bet_{bet_size}"),
            ]

    def _deal_board_card(self, p0_card: int, p1_card: int,
                         pot: float, history: str, depth: int) -> GameNode:
        """Deal community card as a chance node, then build street 2."""
        chance = ChanceNode(depth=depth)
        remaining = [c for c in LEDUC_CARDS if c != p0_card and c != p1_card]
        prob = 1.0 / len(remaining)

        for board_card in remaining:
            label = _cname(board_card)
            new_history = history + "/" + label + "/"
            child = self._build_betting_round(
                p0_card, p1_card, board_card,
                pot=pot, street=2, bet_size=BET_STREET2,
                history=new_history, num_bets=0, depth=depth + 1,
            )
            child.parent = chance
            chance.outcomes[label] = (prob, child)

        return chance

    def _showdown(self, p0_card: int, p1_card: int,
                  board_card: int | None, pot: float, depth: int) -> TerminalNode:
        """Determine winner at showdown."""
        p0_rank = card_rank(p0_card)
        p1_rank = card_rank(p1_card)
        board_rank = card_rank(board_card) if board_card is not None else -1

        p0_pair = p0_rank == board_rank
        p1_pair = p1_rank == board_rank

        if p0_pair and not p1_pair:
            winner = 0
        elif p1_pair and not p0_pair:
            winner = 1
        elif p0_rank > p1_rank:
            winner = 0
        elif p1_rank > p0_rank:
            winner = 1
        else:
            # True tie
            return TerminalNode(depth=depth, payoffs=(0.0, 0.0), pot=pot,
                                is_showdown=True)

        payoff = pot / 2
        payoffs = (payoff, -payoff) if winner == 0 else (-payoff, payoff)
        return TerminalNode(depth=depth, payoffs=payoffs, pot=pot, is_showdown=True)

    def get_info_set_key(self, player: int, cards: Sequence[int],
                         history: str, board: Sequence[int] = ()) -> str:
        """Info set key: 'Jc|/Ks/:cb' format."""
        card_str = _cname(cards[0])
        board_str = "".join(_cname(c) for c in board) if board else ""
        if board_str:
            return f"{card_str}|{board_str}|{history}"
        return f"{card_str}|{history}"


def _cname(c: int) -> str:
    """Card name like 'Jc', 'Ks'."""
    r = LEDUC_RANK_NAMES[card_rank(c)]
    s = "cdhs"[card_suit(c)]
    return f"{r}{s}"

File: lib/solver/variants/leduc.ts
/**
 * Leduc Hold'em — 6-card, 2-street poker game for multi-round CFR validation.
 *
 * Rules:
 * - 6-card deck: J, Q, K with 2 suits each
 * - Each player antes 1 chip, gets 1 private card
 * - Street 1: betting round (check/bet(2))
 * - 1 community card dealt
 * - Street 2: betting round (check/bet(4))
 * - Showdown: pair (private matches board) beats high card; higher rank wins
 * - Max 2 bets per street (bet + raise)
 */

import { card, cardRank, cardSuit } from "../cards";
import {
  Action,
  makeActionInfo,
  type ActionInfo as ActionInfoType,
  type ActionNode,
  type ChanceNode,
  type GameNode,
  type TerminalNode,
  type TreeBuilder,
} from "../types";

// Leduc uses J=9, Q=10, K=11 with suits 0 and 1
const LEDUC_RANKS = [9, 10, 11];
const LEDUC_SUITS = [0, 1];
const LEDUC_CARDS: number[] = [];
for (const r of LEDUC_RANKS) {
  for (const s of LEDUC_SUITS) {
    LEDUC_CARDS.push(card(r, s));
  }
}
const LEDUC_RANK_NAMES: Record<number, string> = { 9: "J", 10: "Q", 11: "K" };

const ANTE = 1;
const BET_STREET1 = 2;
const BET_STREET2 = 4;
const MAX_BETS_PER_STREET = 2;

function cname(c: number): string {
  const r = LEDUC_RANK_NAMES[cardRank(c)];
  const s = "cdhs"[cardSuit(c)];
  return `${r}${s}`;
}

export class LeducTreeBuilder implements TreeBuilder {
  buildTree(): GameNode {
    const root: ChanceNode = {
      type: "chance",
      outcomes: new Map(),
      depth: 0,
    };

    // Deal 2 private cards (all ordered pairs from 6 cards)
    const deals: [number, number][] = [];
    for (const c1 of LEDUC_CARDS) {
      for (const c2 of LEDUC_CARDS) {
        if (c1 !== c2) deals.push([c1, c2]);
      }
    }
    const prob = 1.0 / deals.length; // 1/30

    for (const [p0Card, p1Card] of deals) {
      const label = `${cname(p0Card)}${cname(p1Card)}`;
      const child = this._buildStreet1(p0Card, p1Card, 1);
      root.outcomes.set(label, { prob, child });
    }

    return root;
  }

  private _buildStreet1(
    p0Card: number,
    p1Card: number,
    depth: number,
  ): GameNode {
    const pot = 2 * ANTE;
    return this._buildBettingRound(
      p0Card,
      p1Card,
      null,
      pot,
      1,
      BET_STREET1,
      "",
      0,
      depth,
    );
  }

  private _buildBettingRound(
    p0Card: number,
    p1Card: number,
    boardCard: number | null,
    pot: number,
    street: number,
    betSize: number,
    history: string,
    numBets: number,
    depth: number,
  ): GameNode {
    // Determine acting player
    const slashIdx = history.lastIndexOf("/");
    const currentRound = slashIdx >= 0 ? history.slice(slashIdx + 1) : history;
    // Filter out card labels within current round
    const actionsInRound = currentRound.replace(/[^cfbk]/g, "").length;
    const player = actionsInRound % 2;

    const playerCard = player === 0 ? p0Card : p1Card;
    const board =
      boardCard !== null ? [boardCard] : [];
    const infoKey = this.getInfoSetKey(player, [playerCard], history, board);

    // Terminal conditions
    if (this._isRoundTerminal(history)) {
      if (history.endsWith("f")) {
        // Previous player folded, current player wins
        const winner = player;
        const payoff = pot / 2;
        const payoffs: [number, number] =
          winner === 0 ? [payoff, -payoff] : [-payoff, payoff];
        return {
          type: "terminal",
          depth,
          payoffs,
          pot,
          isShowdown: false,
        };
      }

      // Round ended with check-check or bet-call
      if (street === 1) {
        return this._dealBoardCard(p0Card, p1Card, pot, history, depth);
      } else {
        return this._showdown(p0Card, p1Card, boardCard, pot, depth);
      }
    }

    // Build action node
    const actions = this._getActions(history, numBets, betSize);
    const node: ActionNode = {
      type: "action",
      depth,
      player,
      actions,
      children: new Map(),
      infoSetKey: infoKey,
      pot,
      stacks: [0, 0],
    };

    for (const action of actions) {
      let newHist: string;
      let newPot: number;
      let newBets: number;

      if (action.label === "fold") {
        newHist = history + "f";
        newPot = pot;
        newBets = numBets;
      } else if (action.label === "check") {
        newHist = history + "c";
        newPot = pot;
        newBets = numBets;
      } else if (
        action.label.startsWith("bet") ||
        action.label.startsWith("raise")
      ) {
        newHist = history + "b";
        newPot = pot + betSize;
        newBets = numBets + 1;
      } else if (action.label === "call") {
        newHist = history + "k"; // 'k' for call to avoid ambiguity
        newPot = pot + betSize;
        newBets = numBets;
      } else {
        throw new Error(`Unknown action: ${action.label}`);
      }

      const child = this._buildBettingRound(
        p0Card,
        p1Card,
        boardCard,
        newPot,
        street,
        betSize,
        newHist,
        newBets,
        depth + 1,
      );
      node.children.set(action.label, child);
    }

    return node;
  }

  private _isRoundTerminal(history: string): boolean {
    const slashIdx = history.lastIndexOf("/");
    const current = slashIdx >= 0 ? history.slice(slashIdx + 1) : history;
    // Filter to just action chars
    const acts = current.replace(/[^cfbk]/g, "");
    if (acts.length < 2) return false;
    const lastTwo = acts.slice(-2);
    return lastTwo === "cc" || lastTwo === "bk" || lastTwo === "bf";
  }

  private _getActions(
    history: string,
    numBets: number,
    betSize: number,
  ): ActionInfoType[] {
    const slashIdx = history.lastIndexOf("/");
    const current = slashIdx >= 0 ? history.slice(slashIdx + 1) : history;
    const acts = current.replace(/[^cfbk]/g, "");
    const lastAction = acts.length > 0 ? acts[acts.length - 1] : "";

    if (lastAction === "b") {
      const actions: ActionInfoType[] = [
        makeActionInfo(Action.FOLD, 0, "fold"),
        makeActionInfo(Action.CALL, betSize, "call"),
      ];
      if (numBets < MAX_BETS_PER_STREET) {
        actions.push(makeActionInfo(Action.RAISE, betSize, "raise"));
      }
      return actions;
    } else {
      return [
        makeActionInfo(Action.CHECK, 0, "check"),
        makeActionInfo(Action.BET, betSize, `bet_${betSize}`),
      ];
    }
  }

  private _dealBoardCard(
    p0Card: number,
    p1Card: number,
    pot: number,
    history: string,
    depth: number,
  ): GameNode {
    const chance: ChanceNode = {
      type: "chance",
      outcomes: new Map(),
      depth,
    };
    const remaining = LEDUC_CARDS.filter(
      (c) => c !== p0Card && c !== p1Card,
    );
    const prob = 1.0 / remaining.length;

    for (const boardCard of remaining) {
      const label = cname(boardCard);
      const newHistory = history + "/" + label + "/";
      const child = this._buildBettingRound(
        p0Card,
        p1Card,
        boardCard,
        pot,
        2,
        BET_STREET2,
        newHistory,
        0,
        depth + 1,
      );
      chance.outcomes.set(label, { prob, child });
    }

    return chance;
  }

  private _showdown(
    p0Card: number,
    p1Card: number,
    boardCard: number | null,
    pot: number,
    depth: number,
  ): TerminalNode {
    const p0Rank = cardRank(p0Card);
    const p1Rank = cardRank(p1Card);
    const boardRank = boardCard !== null ? cardRank(boardCard) : -1;

    const p0Pair = p0Rank === boardRank;
    const p1Pair = p1Rank === boardRank;

    let winner: number;
    if (p0Pair && !p1Pair) {
      winner = 0;
    } else if (p1Pair && !p0Pair) {
      winner = 1;
    } else if (p0Rank > p1Rank) {
      winner = 0;
    } else if (p1Rank > p0Rank) {
      winner = 1;
    } else {
      // True tie
      return {
        type: "terminal",
        depth,
        payoffs: [0.0, 0.0],
        pot,
        isShowdown: true,
      };
    }

    const payoff = pot / 2;
    const payoffs: [number, number] =
      winner === 0 ? [payoff, -payoff] : [-payoff, payoff];
    return {
      type: "terminal",
      depth,
      payoffs,
      pot,
      isShowdown: true,
    };
  }

  getInfoSetKey(
    player: number,
    cards: number[],
    history: string,
    board?: number[],
  ): string {
    const cardStr = cname(cards[0]);
    const boardStr = board && board.length > 0 ? board.map(cname).join("") : "";
    if (boardStr) return `${cardStr}|${boardStr}|${history}`;
    return `${cardStr}|${history}`;
  }
}
